# 📊Project: Railway Ticketing | Data Cleaning 

## Introduction

In this project, i will take the role of a **Transportation Data Analyst** to explore a real-world datasets from UK. The main objectives are to **clean the data**, **handle null values**, **update data types**, and **detect anomalies** in the data before moving on to visualization.

In [1]:
#Importing libraries
import pandas as pd
import numpy as np
from datetime import datetime as dt

## Table of Contents (TOCs)

1. [Project](#)
    * [Introduction](#Introduction)
    * [Data Description](#Data-Description)
2. [Data Preparation](#Data-Preparation)
3. [Data Cleaning](#Data-Cleaning)
    * [Data Quality Report](#Data-Quality-Report)
    * [Handling Null Values](#Handling-Null-Values)
    * [Converting Data Types](#Converting-Data-Types)
    * [Creating Column](#Creating-column)
        * [A. "transaction_id" column](#A.-"transaction_id"-column)
        * [B. "arrival_date" column](#B.-"arrival_date"-column)
    * [Renaming Columns](#Renaming-Columns)
    * [Updating Values](#Updating-Values)
        * [A. Update Values in delay_reason columns](#A.-Update-Values-in-delay_reason-columns)
    * [Anomaly Detection](#Anomaly-Detection)
        * [A. Check whether the purchase date after the journey date?](#A.-Check-whether-the-purchase-date-after-the-journey-date-?)
        * [B. Check Whether Ticket Price equal or smaller than 0?](#B.-Check-Whether-Ticket-Price-equal-or-smaller-than-0-?)
4. [Dataset Download](#Dataset-Download)

## Data Description

- **`Transaction ID`**: Unique identifier for an individual train ticket purchase
- **`Date of Purchase`**: Date the ticket was purchased
- **`Time of Purchase`**: Time the ticket was purchased
- **`Purchase Type`**: Whether the ticket was purchased online or directly at a train station
- **`Payment Method`**: Payment method used to purchase the ticket (Contactles Credit Card or Debit Card)
- **`Railcard`**: Whether the passenger is a National Railcard holder (Adult Senior or Disabled) or not (None). Railcard holders get 1/3 off their ticket purchases.
- **`Ticket Class`**: Seat class for the ticket (Standard or First)
- **`Ticket Type`**: When you bought or can use the ticket.
    - **Advance tickets** are 1/2 off and must be purchased at least a day prior to departure. 
    - **Off-Peak tickets** are 1/4 off and must be used outside of peak hours - (weekdays between 6-8am and 4-6pm).
    - **Anytime tickets** are full price and can be bought and used at any time during the day.
- **`Price`**: Final cost of the ticket
- **`Departure Station`**: Station to board the train
- **`Arrival Destination`**: Station to exit the train
- **`Date of Journey`**: Date the train departed
- **`Departure Time`**: Time the train departed
- **`Arrival Time`**: Time the train was scheduled to arrive at its destination (can be on the day after departure)
- **`Actual Arrival Time`**: Time the train arrived at its destination (can be on the day after departure)
- **`Journey Status`**: Whether the train was on time delayed or cancelled
- **`Reason for Delay`**: Reason for the delay or cancellation
- **`Refund Request`**: Whether the passenger requested a refund after a delay or cancellation

# Data Preparation

- **National Rail UK Train Ticket Dataset**: https://www.kaggle.com/datasets/motsimaslam/national-rail-uk-train-ticket-data

(The dataset from 12/2023 -> 04/2024)

In [2]:
railway_ticket = pd.read_csv('railway.csv')

In [3]:
display(railway_ticket)

Transaction ID Date of Purchase Time of Purchase  \
0      da8a6ba8-b3dc-4677-b176       2023-12-08         12:41:11   
1      b0cdd1b0-f214-4197-be53       2023-12-16         11:23:01   
2      f3ba7a96-f713-40d9-9629       2023-12-19         19:51:27   
3      b2471f11-4fe7-4c87-8ab4       2023-12-20         23:00:36   
4      2be00b45-0762-485e-a7a3       2023-12-27         18:22:56   
...                        ...              ...              ...   
31648  1304623d-b8b7-4999-8e9c       2024-04-30         18:42:58   
31649  7da22246-f480-417c-bc2f       2024-04-30         18:46:10   
31650  add9debf-46c1-4c75-b52d       2024-04-30         18:56:41   
31651  b92b047c-21fd-4859-966a       2024-04-30         19:51:47   
31652  1d5d89a2-bde5-410f-8f91       2024-04-30         20:05:39   

      Purchase Type Payment Method Railcard Ticket Class Ticket Type  Price  \
0            Online    Contactless    Adult     Standard     Advance     43   
1           Station    Credit Card    Adult     Standard     Advance     23   
2            Online    Credit Card      NaN     Standard     Advance      3   
3           Station    Credit Card      NaN     Standard     Advance     13   
4            Online    Contactless      NaN     Standard     Advance     76   
...             ...            ...      ...          ...         ...    ...   
31648        Online    Credit Card      NaN     Standard    Off-Peak      4   
31649        Online    Contactless      NaN     Standard    Off-Peak     10   
31650       Station    Credit Card      NaN     Standard    Off-Peak      4   
31651       Station    Credit Card      NaN     Standard    Off-Peak     10   
31652       Station    Credit Card    Adult     Standard    Off-Peak      3   

           Departure Station    Arrival Destination Date of Journey  \
0          London Paddington  Liverpool Lime Street      2024-01-01   
1         London Kings Cross                   York      2024-01-01   
2      Liverpool Lime Street  Manchester Piccadilly      2024-01-02   
3          London Paddington                Reading      2024-01-01   
4      Liverpool Lime Street          London Euston      2024-01-01   
...                      ...                    ...             ...   
31648  Manchester Piccadilly  Liverpool Lime Street      2024-04-30   
31649          London Euston  Birmingham New Street      2024-04-30   
31650  Manchester Piccadilly  Liverpool Lime Street      2024-04-30   
31651          London Euston  Birmingham New Street      2024-04-30   
31652  Liverpool Lime Street  Manchester Piccadilly      2024-04-30   

      Departure Time Arrival Time Actual Arrival Time Journey Status  \
0           11:00:00     13:30:00            13:30:00        On Time   
1           09:45:00     11:35:00            11:40:00        Delayed   
2           18:15:00     18:45:00            18:45:00        On Time   
3           21:30:00     22:30:00            22:30:00        On Time   
4           16:45:00     19:00:00            19:00:00        On Time   
...              ...          ...                 ...            ...   
31648       20:00:00     20:30:00            20:30:00        On Time   
31649       20:15:00     21:35:00            21:35:00        On Time   
31650       20:15:00     20:45:00            20:45:00        On Time   
31651       21:15:00     22:35:00            22:35:00        On Time   
31652       21:30:00     22:00:00            22:00:00        On Time   

      Reason for Delay Refund Request  
0                  NaN             No  
1       Signal Failure             No  
2                  NaN             No  
3                  NaN             No  
4                  NaN             No  
...                ...            ...  
31648              NaN             No  
31649              NaN             No  
31650              NaN             No  
31651              NaN             No  
31652              NaN             No  

[31653 rows x 18 columns]

# Data Cleaning

## Data Quality Report

In [4]:
def check_data_quality(df):
    quality_report = {
        'null_values': df.isnull().sum().to_dict(),
        'duplicate_values': df.duplicated().sum(),
        'info_values': df.info()
    }
    return quality_report

In [5]:
display(check_data_quality(railway_ticket))

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31653 entries, 0 to 31652
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   Transaction ID       31653 non-null  object
 1   Date of Purchase     31653 non-null  object
 2   Time of Purchase     31653 non-null  object
 3   Purchase Type        31653 non-null  object
 4   Payment Method       31653 non-null  object
 5   Railcard             10735 non-null  object
 6   Ticket Class         31653 non-null  object
 7   Ticket Type          31653 non-null  object
 8   Price                31653 non-null  int64 
 9   Departure Station    31653 non-null  object
 10  Arrival Destination  31653 non-null  object
 11  Date of Journey      31653 non-null  object
 12  Departure Time       31653 non-null  object
 13  Arrival Time         31653 non-null  object
 14  Actual Arrival Time  29773 non-null  object
 15  Journey Status       31653 non-null  object
 16  Reas

{'null_values': {'Transaction ID': 0,
  'Date of Purchase': 0,
  'Time of Purchase': 0,
  'Purchase Type': 0,
  'Payment Method': 0,
  'Railcard': 20918,
  'Ticket Class': 0,
  'Ticket Type': 0,
  'Price': 0,
  'Departure Station': 0,
  'Arrival Destination': 0,
  'Date of Journey': 0,
  'Departure Time': 0,
  'Arrival Time': 0,
  'Actual Arrival Time': 1880,
  'Journey Status': 0,
  'Reason for Delay': 27481,
  'Refund Request': 0},
 'duplicate_values': 0,
 'info_values': None}

In [6]:
# Percentage of null values
pct_null_values = (railway_ticket.isnull().sum() * 100 / len(railway_ticket)).sort_values(ascending=False)
pct_null_values.loc[pct_null_values > 0]

Reason for Delay       86.819575
Railcard               66.085363
Actual Arrival Time     5.939405
dtype: float64

Let's take a look:
- The dataset includes **31,652** rows and **18** columns. There are no duplicate ticket transactions, but several columns contains **null values**:
    - `Reason for Delay` and `Railcard` have the highest missing values (more than **50%**).
    - `Actual Arrival Time` has **1,880** rows accounting for **6%** of the dataset.

## Handling Null Values

In [7]:
# "Reason for Delay" column
railway_ticket['Reason for Delay'] = np.where(railway_ticket["Reason for Delay"].isnull(),"None",railway_ticket["Reason for Delay"])

In [8]:
# "Railcard" column
railway_ticket['Railcard'] = np.where(railway_ticket["Railcard"].isnull(),"None", railway_ticket["Railcard"])

- `Reason for Delay` is null because the train journey was not delayed, therefore there is no reason for recorded.
- `Railcard` is null because these passenger are not **National Railcard holders**.

=> We will update both column to **None**

In [9]:
railway_ticket.loc[railway_ticket['Actual Arrival Time'].isnull()]

Transaction ID Date of Purchase Time of Purchase  \
39     69eaff73-ca3a-4831-905b       2024-01-01         03:52:29   
75     a88f097f-bda3-4394-9c7d       2024-01-01         17:43:02   
91     8a66ead7-e381-4311-b667       2024-01-02         00:54:33   
111    d618c56f-3512-484d-9876       2024-01-02         03:33:32   
148    eace2ada-5294-47e6-b4b8       2024-01-02         06:33:00   
...                        ...              ...              ...   
31639  465e3643-fb67-4deb-8ec9       2024-04-30         17:13:32   
31641  2e96cde6-50c1-4311-8089       2024-04-30         17:16:04   
31642  09e207d9-db49-4ec0-9c4f       2024-04-30         17:21:14   
31644  1eacd955-f539-4c79-adb6       2024-04-30         17:26:33   
31645  a0e238b7-1ad1-4bf7-bd6c       2024-04-30         18:11:50   

      Purchase Type Payment Method  Railcard Ticket Class Ticket Type  Price  \
39           Online    Credit Card      None     Standard     Advance      7   
75          Station     Debit Card  Disabled  First Class     Advance     35   
91           Online    Credit Card      None     Standard    Off-Peak     10   
111         Station    Credit Card      None     Standard    Off-Peak     12   
148         Station    Credit Card      None  First Class     Anytime    235   
...             ...            ...       ...          ...         ...    ...   
31639       Station     Debit Card    Senior  First Class     Anytime    144   
31641       Station    Credit Card     Adult     Standard     Anytime     95   
31642        Online    Contactless      None     Standard     Anytime      5   
31644        Online    Credit Card      None     Standard     Anytime     16   
31645        Online    Credit Card    Senior  First Class     Anytime     13   

           Departure Station    Arrival Destination Date of Journey  \
39             London Euston  Birmingham New Street      2024-01-02   
75             London Euston  Birmingham New Street      2024-01-02   
91             London Euston  Birmingham New Street      2024-01-02   
111        London St Pancras  Birmingham New Street      2024-01-02   
148    Liverpool Lime Street          London Euston      2024-01-02   
...                      ...                    ...             ...   
31639          London Euston  Manchester Piccadilly      2024-04-30   
31641          London Euston  Manchester Piccadilly      2024-04-30   
31642  Manchester Piccadilly  Liverpool Lime Street      2024-04-30   
31644      London St Pancras  Birmingham New Street      2024-04-30   
31645  Manchester Piccadilly  Liverpool Lime Street      2024-04-30   

      Departure Time Arrival Time Actual Arrival Time Journey Status  \
39          02:15:00     03:35:00                 NaN      Cancelled   
75          16:00:00     17:20:00                 NaN      Cancelled   
91          02:15:00     03:35:00                 NaN      Cancelled   
111         05:00:00     06:20:00                 NaN      Cancelled   
148         08:00:00     10:15:00                 NaN      Cancelled   
...              ...          ...                 ...            ...   
31639       18:45:00     20:35:00                 NaN      Cancelled   
31641       18:45:00     20:35:00                 NaN      Cancelled   
31642       18:45:00     19:15:00                 NaN      Cancelled   
31644       18:45:00     20:05:00                 NaN      Cancelled   
31645       18:45:00     19:15:00                 NaN      Cancelled   

         Reason for Delay Refund Request  
39        Technical Issue             No  
75        Technical Issue             No  
91        Technical Issue            Yes  
111              Staffing            Yes  
148        Staff Shortage             No  
...                   ...            ...  
31639      Signal failure             No  
31641      Signal failure            Yes  
31642  Weather Conditions             No  
31644      Signal Failure             No  
31645  Weather Conditions             No  

[18

In [10]:
# "Actual Arrival Time" column
x = railway_ticket.loc[railway_ticket['Actual Arrival Time'].isnull()]
np.unique(x['Reason for Delay'])  

array(['Signal Failure', 'Signal failure', 'Staff Shortage', 'Staffing',
       'Technical Issue', 'Traffic', 'Weather', 'Weather Conditions'],
      dtype=object)

`Actual Arrival Time` is null because the train was cancelled => We don't need to update these null values

## Converting Data Types

In [11]:
def converting_date(dataset,col_name):
    dataset[col_name] = pd.to_datetime(dataset[col_name],format = "%Y-%m-%d")

In [12]:
converting_date(railway_ticket,'Date of Purchase')
converting_date(railway_ticket,'Date of Journey')

In [13]:
def converting_data_type(dataset,col_name,dtypes):
    dataset[col_name] = dataset[col_name].astype(dtypes)

In [14]:
converting_data_type(railway_ticket,'Price','Float64')

In [15]:
railway_ticket.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 31653 entries, 0 to 31652
Data columns (total 18 columns):
 #   Column               Non-Null Count  Dtype         
---  ------               --------------  -----         
 0   Transaction ID       31653 non-null  object        
 1   Date of Purchase     31653 non-null  datetime64[ns]
 2   Time of Purchase     31653 non-null  object        
 3   Purchase Type        31653 non-null  object        
 4   Payment Method       31653 non-null  object        
 5   Railcard             31653 non-null  object        
 6   Ticket Class         31653 non-null  object        
 7   Ticket Type          31653 non-null  object        
 8   Price                31653 non-null  Float64       
 9   Departure Station    31653 non-null  object        
 10  Arrival Destination  31653 non-null  object        
 11  Date of Journey      31653 non-null  datetime64[ns]
 12  Departure Time       31653 non-null  object        
 13  Arrival Time         31653 non-

## Creating column

### A. "transaction_id" column

After confirming that there are no duplicate transactions, we will drop the old `TransactionID` column and create a new ID that is easier to read.

In [16]:
railway_ticket  = railway_ticket.drop('Transaction ID', axis = 1)

In [17]:
railway_ticket = railway_ticket.reset_index()

In [18]:
display(railway_ticket)

index Date of Purchase Time of Purchase Purchase Type Payment Method  \
0          0       2023-12-08         12:41:11        Online    Contactless   
1          1       2023-12-16         11:23:01       Station    Credit Card   
2          2       2023-12-19         19:51:27        Online    Credit Card   
3          3       2023-12-20         23:00:36       Station    Credit Card   
4          4       2023-12-27         18:22:56        Online    Contactless   
...      ...              ...              ...           ...            ...   
31648  31648       2024-04-30         18:42:58        Online    Credit Card   
31649  31649       2024-04-30         18:46:10        Online    Contactless   
31650  31650       2024-04-30         18:56:41       Station    Credit Card   
31651  31651       2024-04-30         19:51:47       Station    Credit Card   
31652  31652       2024-04-30         20:05:39       Station    Credit Card   

      Railcard Ticket Class Ticket Type  Price      Departure Station  \
0        Adult     Standard     Advance   43.0      London Paddington   
1        Adult     Standard     Advance   23.0     London Kings Cross   
2         None     Standard     Advance    3.0  Liverpool Lime Street   
3         None     Standard     Advance   13.0      London Paddington   
4         None     Standard     Advance   76.0  Liverpool Lime Street   
...        ...          ...         ...    ...                    ...   
31648     None     Standard    Off-Peak    4.0  Manchester Piccadilly   
31649     None     Standard    Off-Peak   10.0          London Euston   
31650     None     Standard    Off-Peak    4.0  Manchester Piccadilly   
31651     None     Standard    Off-Peak   10.0          London Euston   
31652    Adult     Standard    Off-Peak    3.0  Liverpool Lime Street   

         Arrival Destination Date of Journey Departure Time Arrival Time  \
0      Liverpool Lime Street      2024-01-01       11:00:00     13:30:00   
1                       York      2024-01-01       09:45:00     11:35:00   
2      Manchester Piccadilly      2024-01-02       18:15:00     18:45:00   
3                    Reading      2024-01-01       21:30:00     22:30:00   
4              London Euston      2024-01-01       16:45:00     19:00:00   
...                      ...             ...            ...          ...   
31648  Liverpool Lime Street      2024-04-30       20:00:00     20:30:00   
31649  Birmingham New Street      2024-04-30       20:15:00     21:35:00   
31650  Liverpool Lime Street      2024-04-30       20:15:00     20:45:00   
31651  Birmingham New Street      2024-04-30       21:15:00     22:35:00   
31652  Manchester Piccadilly      2024-04-30       21:30:00     22:00:00   

      Actual Arrival Time Journey Status Reason for Delay Refund Request  
0                13:30:00        On Time             None             No  
1                11:40:00        Delayed   Signal Failure             No  
2                18:45:00        On Time             None             No  
3                22:30:00        On Time             None             No  
4                19:00:00        On Time             None             No  
...                   ...            ...              ...            ...  
31648            20:30:00        On Time             None             No  
31649            21:35:00        On Time             None             No  
31650            20:45:00        On Time             None             No  
31651            22:35:00        On Time             None             No  
31652            22:00:00        On Time             None             No  

[31653 rows x 18 columns]

### B. "arrival_date" column 

Identified **916** cases where the `Actual Arrival Time` < the `Departure Time`, these are **overnight train**

In [19]:
railway_ticket.loc[:,["Date of Journey", "Departure Time", "Actual Arrival Time"]]

Date of Journey Departure Time Actual Arrival Time
0          2024-01-01       11:00:00            13:30:00
1          2024-01-01       09:45:00            11:40:00
2          2024-01-02       18:15:00            18:45:00
3          2024-01-01       21:30:00            22:30:00
4          2024-01-01       16:45:00            19:00:00
...               ...            ...                 ...
31648      2024-04-30       20:00:00            20:30:00
31649      2024-04-30       20:15:00            21:35:00
31650      2024-04-30       20:15:00            20:45:00
31651      2024-04-30       21:15:00            22:35:00
31652      2024-04-30       21:30:00            22:00:00

[31653 rows x 3 columns]

Conditions:
- If **Actual Arrival Time** > **Departure Time** => Keep the `Date of Journey` as the `Date of Arrival`
- If **Actual Arrival Time** < **Departure Time** => `Date of Arrival` = `Date of Journey` + 1 day

In [20]:
railway_ticket.loc[railway_ticket["Actual Arrival Time"] < railway_ticket["Departure Time"],:].head()

index Date of Purchase Time of Purchase Purchase Type Payment Method  \
34     34       2024-01-01         01:10:20       Station    Contactless   
92     92       2024-01-02         01:01:01        Online    Credit Card   
94     94       2024-01-02         01:17:27        Online    Contactless   
95     95       2024-01-02         01:19:18        Online    Contactless   
97     97       2024-01-02         01:26:15        Online    Contactless   

    Railcard Ticket Class Ticket Type  Price      Departure Station  \
34      None     Standard    Off-Peak   10.0          London Euston   
92      None     Standard     Advance    3.0  Liverpool Lime Street   
94  Disabled     Standard     Advance   23.0     London Kings Cross   
95  Disabled     Standard     Advance    5.0                   York   
97  Disabled     Standard     Advance    5.0                   York   

      Arrival Destination Date of Journey Departure Time Arrival Time  \
34  Birmingham New Street      2024-01-01       23:30:00     00:50:00   
92  Manchester Piccadilly      2024-01-03       23:30:00     00:00:00   
94                   York      2024-01-03       23:45:00     01:35:00   
95                 Durham      2024-01-03       23:45:00     00:35:00   
97                 Durham      2024-01-03       23:45:00     00:35:00   

   Actual Arrival Time Journey Status Reason for Delay Refund Request  
34            00:50:00        On Time             None             No  
92            00:00:00        On Time             None             No  
94            01:35:00        On Time             None             No  
95            00:35:00        On Time             None             No  
97            00:35:00        On Time             None             No

In [21]:
# Điều kiện
condition = railway_ticket["Departure Time"] > railway_ticket["Actual Arrival Time"]

# Nhập dữ liệu để Arrival = Journey, sau đó mới bắt đầu +1 ngày với các rows thỏa điều kiện
railway_ticket["Date of Arrival"] = railway_ticket["Date of Journey"]

# Nếu thỏa điều kiện = + thêm 1 ngày => Arrival Date = Journey + 1 day
railway_ticket.loc[condition,"Date of Arrival"] = railway_ticket.loc[condition,"Date of Journey"] + pd.Timedelta(days=1)

In [22]:
railway_ticket.loc[condition,:].head()

index Date of Purchase Time of Purchase Purchase Type Payment Method  \
34     34       2024-01-01         01:10:20       Station    Contactless   
92     92       2024-01-02         01:01:01        Online    Credit Card   
94     94       2024-01-02         01:17:27        Online    Contactless   
95     95       2024-01-02         01:19:18        Online    Contactless   
97     97       2024-01-02         01:26:15        Online    Contactless   

    Railcard Ticket Class Ticket Type  Price      Departure Station  \
34      None     Standard    Off-Peak   10.0          London Euston   
92      None     Standard     Advance    3.0  Liverpool Lime Street   
94  Disabled     Standard     Advance   23.0     London Kings Cross   
95  Disabled     Standard     Advance    5.0                   York   
97  Disabled     Standard     Advance    5.0                   York   

      Arrival Destination Date of Journey Departure Time Arrival Time  \
34  Birmingham New Street      2024-01-01       23:30:00     00:50:00   
92  Manchester Piccadilly      2024-01-03       23:30:00     00:00:00   
94                   York      2024-01-03       23:45:00     01:35:00   
95                 Durham      2024-01-03       23:45:00     00:35:00   
97                 Durham      2024-01-03       23:45:00     00:35:00   

   Actual Arrival Time Journey Status Reason for Delay Refund Request  \
34            00:50:00        On Time             None             No   
92            00:00:00        On Time             None             No   
94            01:35:00        On Time             None             No   
95            00:35:00        On Time             None             No   
97            00:35:00        On Time             None             No   

   Date of Arrival  
34      2024-01-02  
92      2024-01-04  
94      2024-01-04  
95      2024-01-04  
97      2024-01-04

=> Expected Result: after applying this logic, each rows guarantee that: `Departure Time` < `Actual Arrival Time` => Logical timeline

## Renaming Columns

In [23]:
railway_ticket.rename({
    "index": "transaction_id",
    "Date of Purchase": "purchase_date",
    "Time of Purchase": "purchase_time",
    "Purchase Type": "purchase_type",
    "Payment Method": "payment_method",
    "Railcard": "railcard",
    "Ticket Class": "ticket_class",
    "Ticket Type": "ticket_type",
    "Price": "ticket_price",
    "Departure Station": "departure_station",
    "Arrival Destination": "arrival_station",
    "Date of Journey": "journey_date",
    "Departure Time": "departure_time",
    "Date of Arrival": "arrival_date",
    "Arrival Time": "arrival_time",
    "Actual Arrival Time": "actual_arrival_time",
    "Journey Status": "journey_status",
    "Reason for Delay": "delay_reason",
    "Refund Request": "refund_requested"
}, axis = 1, inplace = True)

## Updating Values

### A. Update Values in delay_reason columns

In [24]:
railway_ticket['delay_reason'].value_counts()

delay_reason
None                  27481
Weather                 995
Technical Issue         707
Signal Failure          523
Signal failure          447
Staffing                410
Staff Shortage          399
Weather Conditions      377
Traffic                 314
Name: count, dtype: int64

In [25]:
#Update Weather -> Weather Conditions
railway_ticket['delay_reason'] = np.where(railway_ticket['delay_reason'] == 'Weather','Weather Conditions',railway_ticket['delay_reason'])

In [26]:
# Signal Failure = Technical Issue
railway_ticket['delay_reason'] = np.where(railway_ticket['delay_reason'] == 'Signal failure','Signal Failure',railway_ticket['delay_reason'])
railway_ticket['delay_reason'] = np.where(railway_ticket['delay_reason'] == 'Signal Failure','Technical Issue',railway_ticket['delay_reason'])

## Anomaly Detection

### A. Check whether the purchase date after the journey date ?

In [27]:
datediff_range = abs(railway_ticket["journey_date"] - railway_ticket["purchase_date"])

In [28]:
datediff_range.value_counts()

0 days     14092
1 days     13744
2 days       367
3 days       284
4 days       254
7 days       253
5 days       221
8 days       219
6 days       212
10 days      193
11 days      190
9 days       182
14 days      163
12 days      160
13 days      149
15 days      148
16 days      135
17 days      119
18 days      103
19 days       91
20 days       79
21 days       74
22 days       64
23 days       52
25 days       36
24 days       33
27 days       19
26 days       16
28 days        1
Name: count, dtype: int64

All values are **0 days** or **greater than 0 days**. There are no negative values, meaning there is no case the ticket was purchased after the Journey date.

### B. Check Whether Ticket Price equal or smaller than 0 ?

In [29]:
print((railway_ticket["ticket_price"].value_counts(ascending=False).head(20)))

ticket_price
3.0     3468
8.0     2364
13.0    2139
35.0    2129
7.0     2075
4.0     1882
2.0     1861
5.0     1580
10.0    1292
19.0     935
53.0     817
12.0     816
16.0     611
6.0      606
70.0     557
9.0      469
76.0     456
18.0     377
11.0     373
15.0     370
Name: count, dtype: Int64


In [30]:
print(sorted(railway_ticket["ticket_price"].unique()))

[1.0, 2.0, 3.0, 4.0, 5.0, 6.0, 7.0, 8.0, 9.0, 10.0, 11.0, 12.0, 13.0, 14.0, 15.0, 16.0, 17.0, 18.0, 19.0, 20.0, 21.0, 22.0, 23.0, 24.0, 25.0, 26.0, 27.0, 28.0, 29.0, 30.0, 31.0, 32.0, 33.0, 34.0, 35.0, 36.0, 37.0, 38.0, 39.0, 41.0, 42.0, 43.0, 44.0, 45.0, 47.0, 48.0, 49.0, 50.0, 51.0, 52.0, 53.0, 54.0, 55.0, 56.0, 57.0, 59.0, 63.0, 64.0, 65.0, 66.0, 67.0, 69.0, 70.0, 71.0, 72.0, 73.0, 74.0, 76.0, 77.0, 78.0, 80.0, 82.0, 84.0, 86.0, 88.0, 89.0, 90.0, 93.0, 94.0, 95.0, 96.0, 97.0, 101.0, 102.0, 104.0, 106.0, 107.0, 108.0, 109.0, 110.0, 111.0, 112.0, 113.0, 114.0, 116.0, 117.0, 118.0, 119.0, 121.0, 126.0, 128.0, 129.0, 134.0, 135.0, 143.0, 144.0, 146.0, 151.0, 154.0, 157.0, 158.0, 162.0, 168.0, 171.0, 176.0, 178.0, 180.0, 200.0, 203.0, 211.0, 216.0, 235.0, 238.0, 242.0, 267.0]


# Dataset Download

In [31]:
railway_ticket.to_csv('railway_ticket.csv', index=False)